node/core/approval-voting/src/ops.rs

```diff
@@ -0,0 +1,386 @@
+	// by keeping only the metadata about which blocks reference each candidate.
+	let mut visited_candidates = HashMap::new();
+
+	// All the block heights we visited but didn't necessarily delete everything from.
+	let mut visited_heights = HashMap::new();
+
+	let visit_and_remove_block_entry = |
+		block_hash: Hash,
+		overlayed_db: &mut OverlayedBackend<'_, T>,
+		visited_candidates: &mut HashMap<CandidateHash, CandidateEntry>,
+	| -> SubsystemResult<Vec<Hash>> {
+		let block_entry = match overlayed_db.load_block_entry(&block_hash)? {
+			None => return Ok(Vec::new()),
+			Some(b) => b,
+		};
+
+		overlayed_db.delete_block_entry(&block_hash);
+		for &(_, ref candidate_hash) in block_entry.candidates() {
+			let candidate = match visited_candidates.entry(*candidate_hash) {
+				Entry::Occupied(e) => e.into_mut(),
+				Entry::Vacant(e) => {
+					e.insert(match overlayed_db.load_candidate_entry(candidate_hash)? {
+						None => continue, // Should not happen except for corrupt DB
+						Some(c) => c,
+					})
+				}
+			};
+
+			candidate.block_assignments.remove(&block_hash);
+		}
+
+		Ok(block_entry.children)
+	};
+
+	// First visit everything before the height.
+	for i in range.0..canon_number {
+		let at_height = overlay_db.load_blocks_at_height(&i)?;
+		overlay_db.delete_blocks_at_height(i);
+
+		for b in at_height {
+			let _ = visit_and_remove_block_entry(
+				b,
+				overlay_db,
+				&mut visited_candidates,
+			)?;
+		}
+	}
+
+	// Then visit everything at the height.
+	let pruned_branches = {
+		let at_height = overlay_db.load_blocks_at_height(&canon_number)?;
+		overlay_db.delete_blocks_at_height(canon_number);
+
+		// Note that while there may be branches descending from blocks at earlier heights,
+		// we have already covered them by removing everything at earlier heights.
+		let mut pruned_branches = Vec::new();
+
+		for b in at_height {
+			let children = visit_and_remove_block_entry(
+				b,
+				overlay_db,
+				&mut visited_candidates,
+			)?;
+
+			if b != canon_hash {
+				pruned_branches.extend(children);
+			}
+		}
+
+		pruned_branches
+	};
+
+	// Follow all children of non-canonicalized blocks.
+	{
+		let mut frontier: Vec<(BlockNumber, Hash)> = pruned_branches.into_iter().map(|h| (canon_number + 1, h)).collect();
+		while let Some((height, next_child)) = frontier.pop() {
+			let children = visit_and_remove_block_entry(
+				next_child,
+				overlay_db,
+				&mut visited_candidates,
+			)?;
+
+			// extend the frontier of branches to include the given height.
+			frontier.extend(children.into_iter().map(|h| (height + 1, h)));
+
+			// visit the at-height key for this deleted block's height.
+			let at_height = match visited_heights.entry(height) {
+				Entry::Occupied(e) => e.into_mut(),
+				Entry::Vacant(e) => e.insert(overlay_db.load_blocks_at_height(&height)?),
+			};
+			if let Some(i) = at_height.iter().position(|x| x == &next_child) {
+				at_height.remove(i);
+			}
+		}
+	}
+
```

Comment:
:+1: I like the ration of comments and code, keeps a nice balance

Reply:
just a note, that this whole function was extracted from node/core/approval-voting/src/approval_db/v1/mod.rs

Reply:
correct!